### PR TITLE
Bug 1853264: Fix unbound cardinality for MCDRebootErr and MCDPivotErr

### PR DESCRIPTION
**Pull request by sinnykumari**
Metrics shouldn't log failure, that should be done by logs Rename metrics name to match with metric type being used i.e Counter: ssh_accessed->  ssh_accessed_total
mcd_pivot_err -> mcd_pivot_err_total
mcd_reboot_err -> mcd_reboot_err_total

Followup of PR- https://github.com/openshift/machine-config-operator/pull/2394/ Metric types- https://prometheus.io/docs/concepts/metric_types/ Metric naming-  https://prometheus.io/docs/practices/naming/

Co-authored-by: Kirsten Garrison <kikis.github@gmail.com>

### CODE REVIEW

**Comment by sinnykumari** [comment]
This is followup PR of https://github.com/openshift/machine-config-operator/pull/2394 incorporating changes from the PR, rebased and addressing reviewing comments like updating metrics name as per naming guideline.

@jan--f Since you have reviewed original PR, can you please review this follow-up PR :)
/cc @kikisdeliveryservice @jan--f

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (yuqi-zhang) [inline install/0000_90_machine-config-operator_01_prometheus-rules.yaml:76]: Doesn't have to be part of this PR, but is this still triggerable? Pivot doesn't exist as a service anymore

Comment (sinnykumari) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml:76]: We are using this error in context of OSUpdate https://github.com/openshift/machine-config-operator/blob/master/pkg/daemon/update.go#L2217 . Pivot is the term I think we have been using since beginning, maybe we can revisit later if this name sounds confusing

Comment (yuqi-zhang) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml:76]: ack, sorry, I mis-read the code, I think that's fine for now then, thanks!

**Review by jan--f — APPROVED** [review]
Just one comment, mostly due to my own ignorance. As a whole this
/lgtm

File: pkg/daemon/update.go
Comment (sinnykumari) [reply pkg/daemon/update.go]: Thanks for the review and good observation. 
Until 4.11, osImageURL was under our control as it would be populated with what is shipped through OCP release payload. With 4.12 onward where we moved to layering, we allow user to override OSImageURL field with their custom built OSImage. So, this could lead to unbound cardinality. I have updated commit with label removed, should be good now.

File: pkg/daemon/metrics.go
Comment (simonpasquier) [inline pkg/daemon/metrics.go:27]:
```suggestion
			Help: "Total number of successful SSH logins.",
```

File: pkg/daemon/metrics.go
Comment (simonpasquier) [inline pkg/daemon/metrics.go:41]:
```suggestion
			Help: "Total number of errors encountered during pivot.",
```

File: pkg/daemon/metrics.go
Comment (simonpasquier) [inline pkg/daemon/metrics.go:40]:
```suggestion
			Name: "mcd_pivot_errors_total",
```

File: pkg/daemon/metrics.go
Comment (simonpasquier) [inline pkg/daemon/metrics.go:26]:
```suggestion
			Name: "ssh_successful_logins_total",
```

File: pkg/daemon/metrics.go
Comment (simonpasquier) [inline pkg/daemon/metrics.go:61]:
```suggestion
			Name: "mcd_reboots_failed_total",
			Description: "Total number of reboots that failed.",
```

File: pkg/daemon/metrics.go
Comment (simonpasquier) [inline pkg/daemon/metrics.go:59]: You don't need a CounterVec if you have no labels.

```suggestion
	MCDRebootErr = prometheus.NewCounter(
```

File: pkg/daemon/metrics.go
Comment (simonpasquier) [inline pkg/daemon/metrics.go:38]: You don't need a CounterVec if you have no labels.

```suggestion
	MCDPivotErr = prometheus.NewCounter(
```

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (simonpasquier) [inline install/0000_90_machine-config-operator_01_prometheus-rules.yaml:58]: How/when this alert would clear? As the metric is a counter, its value will only reset/disappear when the process exposing it (MCD?) restarts.

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (simonpasquier) [inline install/0000_90_machine-config-operator_01_prometheus-rules.yaml:76]: Same remark here, you'd usually use `rate()` or `increase()` to alert on a counter metric but not having much knowledge about MCD, it's hard for me to be more specific.

File: pkg/daemon/metrics.go
Comment (sinnykumari) [reply pkg/daemon/metrics.go:26]: When a user directly perform ssh, we annotate node with `machineconfiguration.openshift.io/ssh:accessed` , this was done in the beginning of OCP 4 to discourage users from directly doing ssh into nodes and making any changes. I think, keeping metric name ssh_accessed_total would be less confusing as it aligns with why its usage.

Comment (sinnykumari) [reply pkg/daemon/metrics.go:26]: Will update help message accordingly.

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (sinnykumari) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml:76]: behavior should be same as previous. After pivot (OSUpdate), MCD will perform node reboot.

**Comment by sinnykumari** [comment]
Thanks Jan and Simon for the review, updated PR with the review comments.

File: pkg/daemon/metrics.go
Comment (simonpasquier) [reply pkg/daemon/metrics.go:26]: usually a counter metric should be named like `<noun>_total`, not `<adjective>_total` (e.g. `http_requests_total` , not `http_requested_total`). Maybe `ssh_accesses_total`?

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (sinnykumari) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml]: As per further conversation, have updated PR with use of increase()

File: pkg/daemon/metrics.go
Comment (sinnykumari) [reply pkg/daemon/metrics.go]: updated metric name to `ssh_accesses_total`

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (simonpasquier) [inline install/0000_90_machine-config-operator_01_prometheus-rules.yaml:58]: It's usually a good practice to keep only the labels that are relevant for the context. In this example, I assume that the namespace node and pod labels are enough and that other labels like  instance, service, ... don't bring any value.

```suggestion
             max by(namespace, node, pod) increase(mcd_reboots_failed_total[15m]) > 0
```

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (simonpasquier) [inline install/0000_90_machine-config-operator_01_prometheus-rules.yaml:80]: This is nice! Probably worth doing the same for the other alert?

```suggestion
            message: "Error detected in pivot logs on {{ $labels.node }} , upgrade may be blocked. For more details:  oc logs -f -n {{ $labels.namespace }} {{ $labels.pod }} -c machine-config-daemon "
```

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (sinnykumari) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml:56]: Totally agree with you. I was trying to fix the bug with minimal changes. We have a epic around metrics and alert improvement https://issues.redhat.com/browse/MCO-1 . So, maybe we can make these things better then overall.

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (sinnykumari) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml]: great suggestion, fixed.

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml
Comment (simonpasquier) [inline install/0000_90_machine-config-operator_01_prometheus-rules.yaml:58]:
```suggestion
            max by(namespace, node, pod) (increase(mcd_reboots_failed_total[15m])) > 0
```

Comment (sinnykumari) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml]: ah, thought these parenthesis was optional. Fixed.

**Comment by sinnykumari** [comment]
Noting for future reference, prometheus rule related error logs can be seen in openshift-monitoring_prometheus-operator pod logs,  https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3406/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1595510751693377536/artifacts/e2e-gcp-op/gather-extra/artifacts/pods/openshift-monitoring_prometheus-operator-764b67cd89-klbg7_prometheus-operator.log

**Comment by sinnykumari** [comment]
e2e test looks good
@simonpasquier  need lgtm if this looks good now

**Comment by simonpasquier** [comment]
/lgtm
/hold
Looks good from a monitoring standpoint. Putting on hold in case anyone from machine-config-operator needs to chime in

**Comment by sinnykumari** [comment]
Thanks a lot Simon for the thorough review, very much appreciated!

**Review by simonpasquier — COMMENTED** [review]
You probably want to add a `for: 5m` to the alerts. This would avoid the alert firing immediatly in case it takes a bit of time for the reboot to be effective.

**Review by yuqi-zhang — APPROVED** [review]
/hold cancel

Thanks for the work and review! Let's try to merge this and do any follow ups via another PR if needed. Sorry for the delay

**Comment by rioliu-rh** [comment]
pre-merge testing result.

check prometheus rule content
```
oc get prometheusrule/machine-config-daemon -n openshift-machine-config-operator -o yaml | yq -y '.spec.groups[]|select(.name=="mcd-reboot-error")'
name: mcd-reboot-error
rules:
  - alert: MCDRebootError
    annotations:
      message: 'Reboot failed on {{ $labels.node }} , update may be blocked. For more
        details:  oc logs -f -n {{ $labels.namespace }} {{ $labels.pod }} -c machine-config-daemon '
    expr: 'max by(namespace, node, pod) (increase(mcd_reboots_failed_total[15m]))
      > 0

      '
    labels:
      severity: critical

oc get prometheusrule/machine-config-daemon -n openshift-machine-config-operator -o yaml | yq -y '.spec.groups[]|select(.name=="mcd-pivot-error")'
name: mcd-pivot-error
rules:
  - alert: MCDPivotError
    annotations:
      message: 'Error detected in pivot logs on {{ $labels.node }} , upgrade may be
        blocked. For more details:  oc logs -f -n {{ $labels.namespace }} {{ $labels.pod
        }} -c machine-config-daemon '
    expr: 'max by(namespace, node, pod) (increase(mcd_pivot_errors_total[15m])) >
      0

      '
    labels:
      severity: warning
```

check metrics for mcd and ssh

```
oc rsh -n openshift-monitoring prometheus-k8s-0
sh-4.4$
sh-4.4$ curl -s -k  -H "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`" https://172.30.91.3:9001/metrics | grep -v '#' | egrep 'ssh.*total|mcd.*total'
mcd_pivot_errors_total 0
mcd_reboots_failed_total 0
ssh_accesses_total 0
```

**Comment by rioliu-rh** [comment]
cc: @yuqi-zhang

**Comment by yuqi-zhang** [comment]
Thanks @rioliu-rh ! Sorry for the confusion, I will try to label all 4.13+ PRs before lgtm'ing

**Comment by yuqi-zhang** [comment]
/test e2e-gcp-op
/test e2e-agnostic-upgrade

neither seem related

**Pull request merged**
